text/0137-strict-trace-propagation.md

```diff
@@ -0,0 +1,74 @@
+
+We should implement a way to avoid incorrectly continuing such a trace. The easiest way to do this is to restrict trace continuation to happen for the same Organization only.
+
+# Motivation
+
+Today, if some external service (e.g. Zappier) makes an API call to a customers API, and includes a `sentry-trace` header, the API will continue the trace, even if that belongs to a completely different Sentry Org/Project. This is incorrect and that trace _should not_ be continued, but instead a new trace should be started in this case.
+
+# Background
+
+This problem has existed forever basically, but now is the time to finally fix it.
+
+# Supporting Data
+
+Example issue where this happens: https://github.com/getsentry/sentry-javascript/discussions/12774
+
+# Options Considered
+
+This RFC proposes the following Option (A):
+
+## Option A: (Optionally) pass the Org in baggage header & use it to restrict
+
+We should add a new option to all Sentry SDKs, `org`. This optional option takes an org slug, similar to sentry-cli. For example:
+
+```js
+Sentry.init({
+    dsn: 'xxx',
+    org: 'sentry',
+});
+```
+
+When this option is set, it will propagate the org in the baggage entry as `sentry-org=sentry`. By default, nothing is done with this.
+
+In addition to this, users can opt-into strict behavior by setting `strictTracePropagation: true` in their `init` code:
+
+```js
+Sentry.init({
+    dsn: 'xxx',
+    org: 'sentry',
+    strictTracePropagation: true,
+});
+```
+
+When this is enabled, the SDK should not continue any trace that does not continue a matching `sentry-org` baggage entry.
+
+Clarifications on this:
+
+* If `strictTracePropagation: true` is configured, but no `org` is defined, the SDK may warn.
+* If `strictTracePropagation: true` is configured, but an incoming request has no baggage header at all (but it _has_ a `sentry-trace` header), it should _not_ be continued.
+
+For now, this will be opt-in, but SDKs are free to switch the default to `true` in a major release. In onboarding for new projects, we should include `strictTracePropagation: true` by default.
+
+In addition to allowing to manually configure the `org`, SDKs may also infer from other places. For example, if a bundler plugin is used we may infer it from there, if possible. 
+
+We may also adjust the DSN structure (which is currently being reworked anyhow) to also include the DSN. In this case, new DSNs may infer the org from there, ensuring that users do not need to manually set the `org` in addition to `dsn`.
+
+## Option B
+
+Another option could be to configure something similar to `tracePropagationTargets` that leads to the SDK only continuing traces for allow-listed URLs. But this requires configuration by the user (and it may also change over time, ...).
```

Comment:
I would prefer this option. It allows for more granular control and is in line with how other parts of the SDK work.

We could also come up with Option C, which combines A and B. For org wide DSNs, we will already include a service name to the SDK init, so what about we allow a user to choose which of these services should be considered for trace continuation?
The service would then be attached to the baggage as `sentry-service`.

```js
Sentry.init({
    dsn: 'xxx',
    service: 'my-frontend',
});
```

```js
Sentry.init({
    dsn: 'xxx',
    service: 'my-backend',
    tracePropagationOrigins: ['my-frontend'],
});
```

In terms of how the option is evaluated, I propose we stick to  https://develop.sentry.dev/sdk/performance/#sampling, hence this would apply to **3.**. If we get a propagated sampling decision in the `sentry-trace` header (`-1`), we evaluate the service from the baggage against the allow list.

Reply:
I think the service is not ideal here, because it may match over orgs. E.g. think that zappier has a `backend` service and sends their requests with `sentry-service=backend`, and in my app I do allow-list `backend` - this would be matched but is exactly what we want to avoid here.